sherpa/astro/io/pyfits_backend.py

```diff
@@ -540,47 +512,57 @@ def get_image_data(arg, make_copy=False):
-                raise IOErr('badimg', '')
-
-        data['y'] = numpy.asarray(img.data)
-
-        cdeltp = _get_wcs_key(img, 'CDELT1P', 'CDELT2P')
-        crpixp = _get_wcs_key(img, 'CRPIX1P', 'CRPIX2P')
-        crvalp = _get_wcs_key(img, 'CRVAL1P', 'CRVAL2P')
-        cdeltw = _get_wcs_key(img, 'CDELT1', 'CDELT2')
-        crpixw = _get_wcs_key(img, 'CRPIX1', 'CRPIX2')
-        crvalw = _get_wcs_key(img, 'CRVAL1', 'CRVAL2')
-
-        # proper calculation of cdelt wrt PHYSICAL coords
-        if (isinstance(cdeltw, numpy.ndarray)
-                and isinstance(cdeltp, numpy.ndarray)):
-            cdeltw = cdeltw / cdeltp
-
-        # proper calculation of crpix wrt PHYSICAL coords
-        if (isinstance(crpixw, numpy.ndarray)
-                and isinstance(crvalp, numpy.ndarray)
-                and isinstance(cdeltp, numpy.ndarray)
-                and isinstance(crpixp, numpy.ndarray)):
-            crpixw = crvalp + (crpixw - crpixp) * cdeltp
-
-        sky = None
-        if (transformstatus and isinstance(cdeltp, numpy.ndarray)
-                and isinstance(crpixp, numpy.ndarray)
-                and isinstance(crvalp, numpy.ndarray)):
-            sky = WCS('physical', 'LINEAR', crvalp, crpixp, cdeltp)
-
-        eqpos = None
-        if (transformstatus and isinstance(cdeltw, numpy.ndarray)
-                and isinstance(crpixw, numpy.ndarray)
-                and isinstance(crvalw, numpy.ndarray)):
-            eqpos = WCS('world', 'WCS', crvalw, crpixw, cdeltw)
-
-        data['sky'] = sky
-        data['eqpos'] = eqpos
+        # Look for data in the primary or first block.
+        #
+        img = hdus[0]
+        if img.data is None:
+            img = hdus[1]
+            if img.data is None:
+                raise IOErr('badimg', filename)
+
+        if not isinstance(img, (fits.PrimaryHDU, fits.ImageHDU)):
+            raise IOErr("badimg", filename)
+
+        data['y'] = np.asarray(img.data)
+
+        if HAS_TRANSFORM:
+
+            def _get_wcs_key(key, suffix=""):
+                val1 = _try_key(img, f"{key}1{suffix}")
+                if val1 is None:
+                    return None
+
+                val2 = _try_key(img, f"{key}2{suffix}")
+                if val2 is None:
+                    return None
+
+                return np.array([val1, val2], dtype=SherpaFloat)
+
+            cdeltp = _get_wcs_key('CDELT', 'P')
+            crpixp = _get_wcs_key('CRPIX', 'P')
+            crvalp = _get_wcs_key('CRVAL', 'P')
+            cdeltw = _get_wcs_key('CDELT')
+            crpixw = _get_wcs_key('CRPIX')
+            crvalw = _get_wcs_key('CRVAL')
+
+            # proper calculation of cdelt wrt PHYSICAL coords
+            if cdeltw is not None and cdeltp is not None:
+                cdeltw = cdeltw / cdeltp
+
+            # proper calculation of crpix wrt PHYSICAL coords
+            if (crpixw is not None and crvalp is not None and
+                cdeltp is not None and crpixp is not None):
```

Comment:
Unfortunately the values are `Optional[np.ndarray]` which means that we then get errors because of this - e.g.

```
103 >>> from sherpa.astro.io import read_image
104 >>> img = read_image(data_dir + 'image2.fits')
UNEXPECTED EXCEPTION: ValueError('The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()')
Traceback (most recent call last):
  File "/lagado2.real/local/anaconda/envs/sherpa-xspec/lib/python3.11/doctest.py", line 1355, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest sherpa.astro.data[1]>", line 1, in <module>
  File "/lagado.real/sherpa/sherpa-xspec/sherpa/astro/io/__init__.py", line 389, in read_image
    data, filename = backend.get_image_data(arg)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lagado.real/sherpa/sherpa-xspec/sherpa/astro/io/pyfits_backend.py", line 580, in get_image_data
    if None not in [cdeltw, cdeltp]:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Reply:
[this is for #2017 but I wanted to comment here where the issue was raised]